drivers/user_permission_report/app/controllers/user_permission_report/warehouse_reports/reports_controller.rb

```diff
@@ -6,37 +6,56 @@
 
 module UserPermissionReport::WarehouseReports
   class ReportsController < ApplicationController
-
     before_action :set_group_associations
 
     def index
-      if params[:q].present?
-        @users = User.active.
-          text_search(params[:q]).
-          order(:last_name, :first_name).
-          includes(:roles, access_groups: @group_associations).
-          page(params[:page]).per(25)
-      else
-        @users = User.active.order(:last_name, :first_name).includes(:roles, access_groups: @group_associations).page(params[:page]).per(25)
+      respond_to do |format|
+        format.html do
+          if params[:q].present?
+            @users = User.active.
+              text_search(params[:q]).
+              order(:last_name, :first_name).
+              includes(:roles, access_groups: @group_associations).
+              page(params[:page]).per(25)
+          else
+            @users = User.active.order(:last_name, :first_name).includes(:roles, access_groups: @group_associations).page(params[:page]).per(25)
+          end
+        end
+        format.xlsx do
+          @users = User.active.
+            order(:last_name, :first_name).
+            includes(:roles, access_groups: @group_associations)
+          date = Date.current.strftime('%Y-%m-%d')
+          filename = "user-permissions-#{date}.xlsx"
+          headers['Content-Disposition'] = "attachment; filename=#{filename}"
+        end
       end
     end
 
+    def self.all_associations
+      {
+        data_sources: GrdaWarehouse::DataSource,
+        organizations: GrdaWarehouse::Hud::Organization,
+        projects: GrdaWarehouse::Hud::Project,
+        reports: GrdaWarehouse::WarehouseReports::ReportDefinition,
+        cohorts: GrdaWarehouse::Cohort,
+        project_groups: GrdaWarehouse::ProjectGroup,
+      }
+    end
+
     private def set_group_associations
-      @group_associations = [
-        :data_sources,
-        :organizations,
-        :projects,
-        :reports,
-        :cohorts,
-        :project_groups,
-      ]
+      @group_associations = self.class.all_associations.keys
     end
 
     def associated_items(user, meth)
-      unless @group_associations.include? meth
-        return nil
-      end
-      user.access_groups.map{|g| g.public_send(meth)}.flatten.map(&:name)
+      return nil unless @group_associations.include? meth
```

Comment:
This is fine as written, but I've found doing the following to create much less dependence on order:

```
private def set_group_associations
       @group_associations ||= self.class.all_associations.keys
end

 def associated_items(user, meth)
       return nil unless group_associations.include?(meth)
...
```

Reply:
✅ We resolved this a different way by moving group_associations to the user model.